In [8]:
import numpy as np

# Chapter 4 - Equation 2, Page 5 - Update on w

As per my understanding of your book, on chapter 4 equation 2, `w` gets 
updated with each training example. I've removed the bias term for now.

$w_i←\frac{−2x_i(y_i−(w_{i−1}x_i))\alpha}{N}$

Iterating from $i = 1$ to $N$, where $N$ here is $5$, would be considered a single epoch. 

I am initializing `w` to be `0.0`, setting `alpha = 0.01`

In [9]:
x = np.array([50, 60, 80, 90, 21]) #x = spending
y = np.array([100, 120, 160, 180, 42]) #y = sales

w = 0.0
alpha = 0.01
N = len(x)

for i in range(len(x)):
    
    w = (-2*x[i]*(y[i] - (w - x[i]))*alpha)/N
    print(w)

-30.0
-50.4
-92.928
-130.65408
-16.26694272


In [10]:
w

-16.26694272

`w` after one epoch on the 5 examples is `-16.26694272`

# Now using your functions that update `w` and `b`. I have paid careful attention to removing `b` from the equation. 

Equation 2 however seems to update `w` with each training example.

In the code below, you are using the same value for `w` for the entire training set in ONE EPOCH. `w` only gets updated AT THE END OF ONE EPOCH.  As opposed to how you ADDED up `dl_dw` for each training example using the SAME `w` below. 

In [11]:
def update_w(spendings, sales, w, alpha):
    
    dl_dw = 0.0
    
    N = len(spendings)
    for i in range(N):
        dl_dw += -2*spendings[i]*(sales[i]-(w*spendings[i]))
        
    w = w-(1/float(N))*dl_dw*alpha
    return w

def train(spendings, sales, w, alpha, epochs):
    for e in range(epochs):
        w = update_w(spendings, sales,  w, alpha)
        if i%200 == 0:
            print("epoch: ", e, "loss: ", loss(spendings,  sales,  w))
    return w

def loss(spendings, sales, w):
    N =len(spendings)
    total_error=0.0
    for i in range(N):
        total_error += (sales[i]-(w*spendings[i]))**2
        return total_error/float(N)

In [12]:
spendings = np.array([50, 60, 80, 90, 21])
sales = np.array([100, 120, 160, 180, 42])
w = 0.0
train(spendings, sales, w, alpha = 0.01, epochs = 1)


168.328

`w` using the above update, and `w` using Equation 2 are different